### DISCUSSION TITLE
Fixing bad sectors

### DISCUSSION THREAD

**Discussion by fplanque** [keirf/greaseweazle#479] [Q&A]
I made a raw .scp file with gw and I can visualize this bad sector in HxCFloppyEmulator:

<img width="1720" alt="2024-08-26_00-29-06" src="https://github.com/user-attachments/assets/29c495af-124a-457a-bb26-066a54fc8142">

I an wondering this:

- Are these dots equivalent to "bits" ?
- Are those 4 in the middle the most likely culprits for the bad sector?
- If I were to try all combinations of forcing them up or down, could there possibly be one combination to fall back on a correct CRC?
- Are there tools to do this? At least to manually edit the .scp file bit by bit?

Thank you.

**Comment by keirf** [comment]
A dot at 4us is equivalent to a `1` bit. A dot at 8us is equivalent to 2-bit sequence `01`. Hence you can build up bit strings with at most one 0 bit between any pair of 1 bits. That's because all FM clock bits are 1s, and there is a clock bit before every data bit.

Your header is good (green) so bad bits there aren't a problem. In fact the area of "badness" in the header area is probably in the gap between header and data, and indicates where the data area has been rewritten. Sector headers never get rewritten, except by whole-track format operations.

The pair of out-of-band fluxes towards the right-hand side of your red box are more likely culprits. One is at around 6.5us and the other around 9.5us.

I don't know of any tool to let you nudge these individual flux values manually!

**Comment by fplanque** [comment]
> I don't know of any tool to let you nudge these individual flux values manually!

I stumbled onto this today: https://info-coach.fr/atari/software/pc-projects/Aufit.php

It is skewed towards Atari ST formats BUT it says it does this:

> The deep analysis of the FD content allows in some cases an automatic correction of errors found during imaging. For example it is sometimes possible to correctly save an image with partially damaged sectors.

**Reply by keirf** [reply]
Great software by DrCoolZic. I don't know if it can do what you want, but it can do plenty else anyway.
